# Домашнее задание №14

## Номер 1

#### Доказать: 
$$\frac{(RSS_{restricted} - RSS_{unrestricted})/q}{RSS_{unrestricted}/(n-k-1)}=
\frac{(R^2_{unrestricted} - R^2_{restricted})/q}{(1-R^2_{unrestricted})/(n-k-1)}
$$

#### Доказательство:
$R^2 = 1 - \frac{RSS}{TSS} \Rightarrow TSS\cdot R^2 = TSS - RSS \Rightarrow 
RSS = TSS(1-R^2)$

т.к. $\ TSS = \sum^n_{i=1}(y_i - \bar{y})^2 \Rightarrow TSS $ не зависит от типа модели и всегда одно и тоже $\Rightarrow$ 

$RSS_{restricted} = TSS(1-R_{restricted}^2)$

$RSS_{unrestricted} = TSS(1-R_{unrestricted}^2)$

$$\frac{(RSS_{restricted} - RSS_{unrestricted})/q}{RSS_{unrestricted}/(n-k-1)}=
\frac{(TSS(1-R_{restricted}^2) - TSS(1-R_{unrestricted}^2))/q}{TSS(1-R_{unrestricted}^2)/(n-k-1)} = $$
$$\frac{(1-R_{restricted}^2 - 1 + R_{unrestricted}^2)/q}{(1-R_{unrestricted}^2)/(n-k-1)}=
\frac{(R^2_{unrestricted} - R^2_{restricted})/q}{(1-R^2_{unrestricted})/(n-k-1)}
$$

## Номер 2

#### Подключение библиотек

In [2]:
from matplotlib import pyplot as plt

from utils import *

import warnings
warnings.filterwarnings('ignore')
%load_ext autoreload
%autoreload 2

#### Определение параметров для модели

In [3]:
dist_params = dict(

    # Зададим параметры распределения факторов
    x1_mean = 5,
    x1_std = 3,
    x2_mean = 9,
    x2_std = 2,
    x3_mean = -5,
    x3_std = 1,
    corr_12 = 0,
    corr_13 = 0,
    corr_23 = 0.8,
    
    
    # Зададим параметры распределения ошибки
    e_mean = 0,
    e_std = 7,

    # Укажем размер выборки
    N = 1000,

    # Зададим действительные параметры модели
    beta0 = 3,
    beta1 = 10.3,
    beta2 = -7.2,
    beta3 = 3.9
)

# Установим стартовую точку для алгоритма генерации случайных чисел
RANDOM_SEED = 42

In [4]:
data = gen_data(y_type='multivariate', params=dist_params, seed=RANDOM_SEED)

display(data)

x1         x2        x3          e          y
0    6.490142   9.132227 -4.551725 -13.354653 -27.009944
1    9.569090   9.517017 -4.917353  -6.022695   7.838730
2    9.737638   7.577840 -5.899533  -2.895239  22.833812
3    6.627680  10.023754 -4.837624  13.213814  -6.558846
4    5.725887  13.183124 -4.229018   3.895872 -45.539155
..        ...        ...       ...        ...        ...
995  2.119861   9.223365 -4.846865   0.199203 -60.277231
996  1.609389   3.872060 -6.306186 -14.544683 -47.440938
997  6.806355   8.903531 -5.169898  -2.242085 -13.404650
998  2.144245   8.789224 -4.935717  11.503647 -45.942344
999  1.274718   8.369898 -5.366705   2.524535 -62.539286

[1000 rows x 5 columns]

In [10]:
data, model_linearity = train_model(data, target='y', feature_names=['x1', 'x2', 'x3'], show_results=True, show_plots=False, pairwise=False, return_norm_tests=True)


                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.951
Model:                            OLS   Adj. R-squared:                  0.951
Method:                 Least Squares   F-statistic:                     6449.
Date:                Sun, 03 Mar 2024   Prob (F-statistic):               0.00
Time:                        18:02:07   Log-Likelihood:                -3389.9
No. Observations:                1000   AIC:                             6788.
Df Residuals:                     996   BIC:                             6807.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.7953      3.471     -0.517      0.6

Попробуем взять только по два признака ($x_1$ и $x_2$) или ($x_1$ и $x_3$):

In [9]:
data, model_linearity = train_model(data, target='y', feature_names=['x1', 'x2'], show_results=True, show_plots=False, pairwise=False, return_norm_tests=True)

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.947
Model:                            OLS   Adj. R-squared:                  0.947
Method:                 Least Squares   F-statistic:                     8918.
Date:                Sun, 03 Mar 2024   Prob (F-statistic):               0.00
Time:                        18:02:03   Log-Likelihood:                -3429.0
No. Observations:                1000   AIC:                             6864.
Df Residuals:                     997   BIC:                             6879.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -31.4294      1.139    -27.584      0.0

In [11]:
data, model_linearity = train_model(data, target='y', feature_names=['x1', 'x3'], show_results=True, show_plots=False, pairwise=False, return_norm_tests=True)


                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.885
Model:                            OLS   Adj. R-squared:                  0.885
Method:                 Least Squares   F-statistic:                     3842.
Date:                Sun, 03 Mar 2024   Prob (F-statistic):               0.00
Time:                        18:03:36   Log-Likelihood:                -3816.2
No. Observations:                1000   AIC:                             7638.
Df Residuals:                     997   BIC:                             7653.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       -120.3576      1.912    -62.962      0.0

#### Вывод

**1. Оценки коэффициентов модели:**
- Коэффициент ​$\beta_0$ оценен как $−1.7953$, что отличается от реального значения $3$. Смещение довольно большое.
- Оценка для ​$\beta_1$ составляет $10.3397$, близкая к реальному значению 
$10.3$, что может указывать на хорошую оценку этого коэффициента.
- Оценка ​$\beta_2$ равна $-6.9298$, тоже близкое к истинному значению $-7.2$.
- Оценка для ​$\beta_3$ равна $3.4923$, что близко к реальному значению $3.9$

**2. Стандартные ошибки, t-статистики, p-value и доверительные интервалы:**
- **Коэффициент ​$\beta_1$:**

> Стандартная ошибка коэффициента составляет 0.078. Это относительно низкое значение, что указывает на высокую точность оценки.

> Значение t-статистики равно 132.087, а p-value равно 0.00, что говорит о статистической значимости коэффициента.

> Доверительный интервал для коэффициента находится в диапазоне от 10.186 до 10.493. Истинное значение коэффициента попадает в интервал.

- **Коэффициент ​$\beta_2$:**

> Стандартная ошибка коэффициента равна 0.189. Относительно низкое значение, что указывает на высокую точность оценки.

> Значение t-статистики составляет -36.608, а p-value равно 0.00, что говорит о статистической значимости коэффициента.

> Доверительный интервал для коэффициента находится в диапазоне от -7.301 до -6.558. Истинное значение коэффициента попадает в интервал.

- **Коэффициент ​$\beta_3$:**

> Стандартная ошибка коэффициента равна 0.388. Это относительно низкое значение, что указывает на высокую точность оценки.

> Значение t-статистики составляет 8.998, а p-value равно 0.00, что говорит о статистической значимости коэффициента.

> Доверительный интервал для коэффициента находится в диапазоне от 2.731 до 4.254. Истинное значение коэффициента попадает в интервал.

**3. F-статистика, критерии AIC и BIC:**
- Значение **F-статистики** равно 6449. Это высокое значение F-статистики говорит о том, что в целом наша модель является статистически значимой. Это подтверждает, что как минимум один из коэффициентов в модели отличен от нуля.
- **AIC и BIC**: AIC составляет 6788, BIC равно 6808. Критерии AIC и BIC оценивают качество модели, учитывая баланс между точностью и сложностью модели. Меньшее значение критерия указывает на более предпочтительную модель, в результате чего модель с тремя параметрами оказалась предпочтительнее моделей с двумя параметрами.
